File: examples/backends/typescript/test/support.ts
/**
 * Shared fixtures for the offline tests.
 *
 * Nothing here opens a socket. The router is a pure function, so the tests call it
 * directly; and the SDK takes a custom `fetch`, so "Meteroid" is a function that either
 * answers from a table or fails the test for having been called at all.
 */

import assert from "node:assert/strict";

import { Currency, Meteroid, Webhook } from "@meteroid/sdk";

import { router } from "../src/app.js";
import type { Config } from "../src/config.js";
import { AppState } from "../src/state.js";

/** Any base64 string is a valid Standard Webhooks secret. */
export const WEBHOOK_SECRET = "c2NyaWJlLWRlbW8td2ViaG9vay1zZWNyZXQ=";
export const SESSION_SECRET = "test-session-secret";

export const CONFIG: Config = {
  // No API key: every Meteroid-backed operation is out of scope unless a test stubs it.
  meteroidApiKey: "",
  meteroidBaseUrl: "http://meteroid.invalid",
  meteroidWebhookSecret: WEBHOOK_SECRET,
  sessionSecret: SESSION_SECRET,
  defaultCurrency: Currency.Usd,
  port: 0,
};

/** One stubbed Meteroid answer: a JSON body, or a function of the request producing one. */
export type Stub = unknown | ((request: { url: URL; body: unknown }) => unknown);

export interface FakeMeteroid {
  client: Meteroid;
  /** Every request the SDK made, as `"METHOD /path"`. */
  calls: string[];
  /** The parsed JSON body of each request, in the same order. */
  bodies: unknown[];
}

/**
 * A Meteroid client whose `fetch` answers from `routes` (keyed `"METHOD /path"`). A
 * request with no stub is a `404` in Meteroid's own error envelope.
 */
export function fakeMeteroid(routes: Record<string, Stub> = {}): FakeMeteroid {
  const calls: string[] = [];
  const bodies: unknown[] = [];

  const fetch: typeof globalThis.fetch = async (input, init) => {
    const url = new URL(String(input));
    const key = `${init?.method ?? "GET"} ${url.pathname}`;
    const body = typeof init?.body === "string" ? (JSON.parse(init.body) as unknown) : undefined;
    calls.push(key);
    bodies.push(body);

    const stub = routes[key];
    if (stub === undefined) {
      return Response.json({ code: "NOT_FOUND", message: `no stub for ${key}` }, { status: 404 });
    }
    return Response.json(typeof stub === "function" ? stub({ url, body }) : stub);
  };

  return {
    client: new Meteroid("test-key", { serverUrl: "http://meteroid.test", fetch, numRetries: 0 }),
    calls,
    bodies,
  };
}

export interface TestResponse {
  status: number;
  headers: Record<string, string>;
  /** `any` on purpose: the tests index into whatever shape the operation answers with. */
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  body: any;
}

export interface SendOptions {
  headers?: Record<string, string>;
  body?: string | Buffer | null;
  meteroid?: Meteroid;
  state?: AppState;
}

export async function send(
  method: string,
  target: string,
  options: SendOptions = {},
): Promise<TestResponse> {
  const state = options.state ?? new AppState(CONFIG, options.meteroid ?? fakeMeteroid().client);
  const url = new URL(target, "http://localhost");
  const body = options.body === undefined ? "" : options.body;

  const response = await router(state)({
    method,
    path: url.pathname,
    query: url.searchParams,
    headers: options.headers ?? {},
    body: body === null ? null : Buffer.from(body),
  });
  return {
    status: response.status,
    headers: response.headers,
    body: response.body === "" ? null : JSON.parse(response.body),
  };
}

/**
 * Every error in the contract is the same envelope, with `quota` and
 * `upgrade_plan_code` always present — never absent, `null` when they do not apply.
 */
export function assertErrorEnvelope(body: unknown, code: string): void {
  assert.ok(typeof body === "object" && body !== null, `not an object: ${JSON.stringify(body)}`);
  assert.deepEqual(Object.keys(body).sort(), ["code", "message", "quota", "upgrade_plan_code"]);
  const envelope = body as Record<string, unknown>;
  assert.equal(envelope["code"], code, JSON.stringify(body));
  assert.equal(typeof envelope["message"], "string");
}

/** Sign a payload exactly the way Meteroid does, using the SDK's own signer. */
export function signed(payload: string, msgId: string, prefix = "webhook"): Record<string, string> {
  const now = new Date();
  return {
    "content-type": "application/json",
    [`${prefix}-id`]: msgId,
    [`${prefix}-timestamp`]: String(Math.floor(now.getTime() / 1000)),
    [`${prefix}-signature`]: new Webhook(WEBHOOK_SECRET).sign(msgId, now, payload),
  };
}

File: examples/backends/python/tests/support.py
"""Shared fixtures for the offline tests.

Nothing here opens a socket. The router is a pure function, so the tests call it
directly; and the SDK takes a custom ``httpx`` client, so "Meteroid" is an
``httpx.MockTransport`` that either answers from a table or reports, in ``calls``, that
it was reached at all.
"""

from __future__ import annotations

import asyncio
import json
import time
from collections.abc import Callable, Mapping
from dataclasses import dataclass, field, replace
from typing import Any
from urllib.parse import parse_qsl, urlsplit

import httpx
from meteroid import MeteroidAsync, MeteroidOptions, Webhook
from meteroid.models import Currency

from scribe.app import router
from scribe.config import Config
from scribe.http import ScribeRequest
from scribe.state import AppState

# Any base64 string is a valid Standard Webhooks secret.
WEBHOOK_SECRET = "c2NyaWJlLWRlbW8td2ViaG9vay1zZWNyZXQ="
SESSION_SECRET = "test-session-secret"

CONFIG = Config(
    # No API key: every Meteroid-backed operation is out of scope unless a test stubs it.
    meteroid_api_key="",
    meteroid_base_url="http://meteroid.invalid",
    meteroid_webhook_secret=WEBHOOK_SECRET,
    session_secret=SESSION_SECRET,
    default_currency=Currency.USD,
    port=0,
)

# One stubbed Meteroid answer: a JSON body, or a function of the request body producing one.
Stub = Any | Callable[[Any], Any]


@dataclass
class FakeMeteroid:
    client: MeteroidAsync
    # Every request the SDK made, as `"METHOD /path"`.
    calls: list[str] = field(default_factory=list)
    # The parsed JSON body of each request, in the same order.
    bodies: list[Any] = field(default_factory=list)


def meteroid_over(handler: Callable[[httpx.Request], httpx.Response]) -> MeteroidAsync:
    """The real SDK client, with `handler` where the network would be."""
    return MeteroidAsync(
        "test-key",
        MeteroidOptions(server_url="http://meteroid.test", num_retries=0),
        httpx.AsyncClient(transport=httpx.MockTransport(handler)),
    )


def fake_meteroid(routes: Mapping[str, Stub] | None = None) -> FakeMeteroid:
    """A Meteroid client that answers from `routes` (keyed `"METHOD /path"`). A request
    with no stub is a `404` in Meteroid's own error envelope.
    """
    stubs = dict(routes or {})

    def handler(request: httpx.Request) -> httpx.Response:
        key = f"{request.method} {request.url.path}"
        body = json.loads(request.content) if request.content else None
        fake.calls.append(key)
        fake.bodies.append(body)

        if key not in stubs:
            return httpx.Response(404, json={"code": "NOT_FOUND", "message": f"no stub for {key}"})
        stub = stubs[key]
        return httpx.Response(200, json=stub(body) if callable(stub) else stub)

    fake = FakeMeteroid(client=meteroid_over(handler))
    return fake


@dataclass
class TestResponse:
    __test__ = False

    status: int
    headers: dict[str, str]
    # `Any` on purpose: the tests index into whatever shape the operation answers with.
    body: Any


def state_with(meteroid: MeteroidAsync | None = None, **overrides: Any) -> AppState:
    return AppState(replace(CONFIG, **overrides), meteroid or fake_meteroid().client)


def send(
    method: str,
    target: str,
    *,
    headers: Mapping[str, str] | None = None,
    body: str | bytes | None = "",
    meteroid: MeteroidAsync | None = None,
    state: AppState | None = None,
) -> TestResponse:
    """Drive the router in-process. `body=None` is how `http.py` reports a body it
    stopped buffering at the limit.
    """
    url = urlsplit(target)
    request = ScribeRequest(
        method=method,
        path=url.path,
        query=parse_qsl(url.query, keep_blank_values=True),
        headers=dict(headers or {}),
        body=body.encode() if isinstance(body, str) else body,
    )
    response = asyncio.run(router(state or state_with(meteroid))(request))
    return TestResponse(
        status=response.status,
        headers=response.headers,
        body=json.loads(response.body) if response.body else None,
    )


def assert_error_envelope(body: Any, code: str) -> None:
    """Every error in the contract is the same envelope, with `quota` and
    `upgrade_plan_code` always present — never absent, `null` when they do not apply.
    """
    assert isinstance(body, dict), body
    assert sorted(body) == ["code", "message", "quota", "upgrade_plan_code"]
    assert body["code"] == code, body
    assert isinstance(body["message"], str)


def signed(payload: str | bytes, msg_id: str, prefix: str = "webhook") -> dict[str, str]:
    """Sign a payload exactly the way Meteroid does, using the SDK's own signer."""
    now = int(time.time())
    return {
        "content-type": "application/json",
        f"{prefix}-id": msg_id,
        f"{prefix}-timestamp": str(now),
        f"{prefix}-signature": Webhook(WEBHOOK_SECRET).sign(msg_id, now, payload),
    }
